src/main/java/org/springframework/data/neo4j/core/TemplateSupport.java

```diff
@@ -414,10 +427,35 @@ static <T> Object retrieveOrSetRelatedId(
 	 * @return {@literal true} if renderer will use elementId
 	 */
 	static boolean rendererCanUseElementIdIfPresent(Renderer renderer, Neo4jPersistentEntity<?> targetEntity) {
-		return !targetEntity.isUsingDeprecatedInternalId() && targetEntity.isUsingInternalIds() && renderer.render(Cypher.returning(Functions.elementId(Cypher.anyNode("n"))).build())
+		return !targetEntity.isUsingDeprecatedInternalId() && targetEntity.isUsingInternalIds() && rendererRendersElementId(renderer);
+	}
+
+	private static boolean rendererRendersElementId(Renderer renderer) {
+		return renderer.render(Cypher.returning(Functions.elementId(Cypher.anyNode("n"))).build())
 				.equals("RETURN elementId(n)");
 	}
 
+	public static String convertIdOrElementIdToString(Object value) {
+		if (value instanceof Value driverValue) {
+			if (driverValue.hasType(TypeSystem.getDefault().NUMBER())) {
+				return driverValue.asNumber().toString();
+			}
+			return driverValue.asString();
+		}
+
+		return value.toString();
+	}
+
+	static Object convertToLongIdOrStringElementId(Collection<String> ids) {
```

Comment:
Why returning an Object?

Reply:
The method takes the elementId _or_ id in `String` format, because it got unified to `String` when we got it from the response. To provide it again in a subsequent query, it has to have the right type again (`Long` or `String`). Basically the callee-side calls this method and expects the "right" object type.

Reply:
Aah yes, it will be the original list in case one fails, either heterogeneous or homogeneous. All good then.